src/lightning/data/streaming/downloader.py

```diff
@@ -54,17 +54,23 @@ def download_file(self, remote_filepath: str, local_filepath: str) -> None:
         self._client.client.download_file(
             obj.netloc,
             obj.path.lstrip("/"),
-            local_filepath,
+            local_filepath + "_tmp",
             ExtraArgs=extra_args,
             Config=TransferConfig(use_threads=False),
         )
 
+        if not os.path.exists(local_filepath):
+            shutil.move(local_filepath + "_tmp", local_filepath)
+        else:
+            shutil.rmtree(local_filepath + "_tmp")
```

Comment:
See comment above, I'm pretty sure since now the shard rank is removed from the directory name, there will be a rare race condition here where multiple ranks will check and move/delete.

Reply:
I added filelock to prevent it. It should be fine now.